File: src/one.rs
use itertools::Itertools;

pub fn solve(input: String) {
    let input: Vec<u32> = parse(input);
    let answer_one = part_one(input.clone());
    println!("Part one: {}", answer_one);
    let answer_two = part_two(input.clone());
    println!("Part one: {}", answer_two);
}

fn parse(input: String) -> Vec<u32> {
    input.lines().map(|n| n.parse::<u32>().unwrap()).collect()
}

fn part_one(input: Vec<u32>) -> u32 {
    *input.iter().permutations(2)
        .map(|v| {if v[0] + v[1] == 2020 { v[0] * v[1] } else { 0 }})
        .filter(|n| *n != 0)
        .collect::<Vec<u32>>()
        .first()
        .unwrap()
}

fn part_two(input: Vec<u32>) -> u32 {
    *input.iter().permutations(3)
        .map(|v| {if v[0] + v[1] + v[2] == 2020 { v[0] * v[1] * v[2] } else { 0 }})
        .filter(|n| *n != 0)
        .collect::<Vec<u32>>()
        .first()
        .unwrap()
}

File: aoc/2020/python/one.py
from itertools import product
from functools import reduce
from operator import mul


class Solver:
    def __init__(self, input_file):
        self.input = self.parse(input_file)

    def parse(self, input_file):
        with open(input_file, 'r') as f:
            data = f.readlines()
        return [int(x.strip()) for x in data]

    def part_one(self):
        i = self.input
        return [reduce(mul, x) for x in product(i, i) if sum(x) == 2020][0]

    def part_two(self):
        i = self.input
        # This is pretty, but slow. :sad_pepe:
        # return [reduce(mul, x) for x in product(i, i, i) if sum(x) == 2020][0]
        # This is fast, but ugly.
        for x in i:
            for y in i:
                for z in i:
                    if x+y+z == 2020:
                        return x*y*z

    def solve(self):
        print(f"Part One: {self.part_one()}")
        s = self.part_two()
        if s is not None:
            print(f"Part Two: {s}")
